File: neel/A_Yet_Another_Coin_Problem.py
PI = 3.14159265358979323846
E = 2.718281828459045
INF = 1000000000000000000
MOD1 = 1000000007
MOD2 = 998244353

coin = [1,3,6,10,15]

dp = [INF]*31
dp[0] = 0

for i in range(1,31):
    for j in range(5):
        if i-coin[j] >= 0:
            dp[i] = min(dp[i],1+dp[i-coin[j]])

def solve():
    n = int(input())
    res = min((n//30)*2+dp[n%30],1+((n-15)//30)*2+dp[(n-15)%30] if n>=15 else INF)
    print(res)

def main():
    t = 1
    t = int(input()) 
    for _ in range(t):
        solve()

main()

File: neel/A_Yet_Another_Coin_Problem.cpp
#include <bits/stdc++.h>
using namespace std;

#define ll long long

const ll INF = 1e18;
const int MOD1 = 1e9 + 7;
const int MOD2 = 998244353;

unordered_map<ll, ll> dp;
vector<int> coin = {1, 3, 6, 10, 15};

ll rec(ll n) {
    if (dp.count(n)) return dp[n];

    ll res = INF;

    for (int i = 0; i < 5; i++) {
        if (n - coin[i] >= 0) {
            res = min(res, 1 + rec(n - coin[i]));
        }
    }

    return dp[n] = res;
}

void solve() {
    ll n;
    cin >> n;
    cout << rec(n) << "\n";
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(NULL);

    int t;
    cin >> t;
    while (t--) {
        solve();
    }

    return 0;
}